numpy/_core/tests/test_indexing.py

```diff
@@ -1434,3 +1418,190 @@ def test_setitem(self):
+
+        indices_2d = np.array([[1, 2], [3, 4]])
+        assert_array_equal(a.flat[indices_2d], indices_2d)
+
+        assert_array_equal(a.flat[[True, 1]], a.flat[[1, 1]])
+
+        with pytest.raises(IndexError, match="boolean index did not match indexed flat iterator"):
+            a.flat[[True, True]]
+
+    def test_flatiter_indexing_not_supported_newaxis_mutlidimensional_float(self):
+        a = np.arange(9).reshape((3, 3))
+        with pytest.raises(IndexError, match=r"only integers, slices \(`:`\), ellipsis \(`\.\.\.`\) "
+                                             r"and integer or boolean arrays are valid indices"):
+            a.flat[None]
+
+        with pytest.raises(IndexError, match=r"too many indices for flat iterator: flat iterator is "
+                                             r"1-dimensional, but 2 were indexed"):
+            a.flat[1, 2]
+
+        with pytest.raises(IndexError, match=r"only integers, slices \(`:`\), ellipsis \(`\.\.\.`\) "
+                                             r"and integer or boolean arrays are valid indices"):
+            a.flat[[1.0, 2.0]]
+
+    def test_flatiter_assign_single_integer(self):
+        a = np.arange(9).reshape((3, 3))
+
+        a.flat[0] = 10
+        assert_array_equal(a, np.array([[10, 1, 2], [3, 4, 5], [6, 7, 8]]))
+
+        a.flat[4] = 20
+        assert_array_equal(a, np.array([[10, 1, 2], [3, 20, 5], [6, 7, 8]]))
+
+        a.flat[-1] = 30
+        assert_array_equal(a, np.array([[10, 1, 2], [3, 20, 5], [6, 7, 30]]))
+
+        with pytest.raises(IndexError, match="index 9 is out of bounds"):
+            a.flat[9] = 40
+
+    def test_flatiter_indexing_slice(self):
+        a = np.arange(9).reshape((3, 3))
+        a.flat[:] = 10
+        assert_array_equal(a, np.full((3, 3), 10))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[:5] = 20
+        assert_array_equal(a, np.array([[20, 20, 20], [20, 20, 5], [6, 7, 8]]))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[5:10] = 30
+        assert_array_equal(a, np.array([[0, 1, 2], [3, 4, 30], [30, 30, 30]]))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[::2] = 40
+        assert_array_equal(a, np.array([[40, 1, 40], [3, 40, 5], [40, 7, 40]]))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[::-1] = 50
+        assert_array_equal(a, np.full((3, 3), 50))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[10:5] = 60
+        assert_array_equal(a, np.arange(9).reshape((3, 3)))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[()] = 70
+        assert_array_equal(a, np.full((3, 3), 70))
+
+        a = np.arange(9).reshape((3, 3))
+        a.flat[...] = 80
+        assert_array_equal(a, np.full((3, 3), 80))
+
+    def test_flatiter_indexing_boolean(self):
+        a = np.arange(9).reshape((3, 3))
+        a.flat[True] = 10
+        assert_array_equal(a, np.array([[10, 1, 2], [3, 4, 5], [6, 7, 8]]))
```

Comment:
Technically wrong. `a.flat[True]` should return the `a.reshape(1, a.size)` effectively.  So if anything it would assign it to everything.

In practice, I may be tempted to just deprecate it, since it seems somewhat useless?